### DISCUSSION TITLE
Improving CI with hooks

### DISCUSSION THREAD

**Discussion by JOTSR** [webui-dev/webui#136] [Q&A]
I am currently exploring pre-commit hooks to format code and keep consistency within the repo.
So I have some questions about the technical choices:
* What type of formatter should I use?
* What config should I choose?
* How do I add a pre-commit hook?

I have made some tests with a local pre-commit hook with the following formatters:
```sh
# Format .js/.ts .md ...
prettier *  --write --single-quote --no-semi --use-tabs
# Format .c/.h
find . -iname *.c -o -name *.h | xargs clang-format --style=llvm -i
```
Are you okay with the formatters and config?

To deal with installing hooks, I thought of using Husky, but maybe there is a more convenient tool for a C project. Do you have any ideas or suggestions?

**Comment by AlbertShown** [comment]
I contributed to the C Core in this project because I like using the web in my GUIs and because the code is formatted with 4 spaces as 1 tab 😆
I'm okay (_C/C++ codes_) as they are formatted in `1 Tab = 4 spaces`.

**Reply by ttytm** [reply]
One tab is always one indentation. Users can configure it's width to their preference and dynamically change it without making a change in the code base. People with impaired sight highly benefit from this increased accessibility. For me it's a lot better developer experience with tabs. Space indentation is for presentation, for alignment basically not for indenting real code imho.

**Reply by AlbertShown** [reply]
I know about the indentation, but I didn't know that after changing the Editor settings, it does not modify the file indentations!
Thanks, Good to know 👍

**Reply by AlbertShown** [reply]
If we agreed on a format, we should apply it to all repos. I guess.

**Reply by JOTSR** [reply]
For the same accessibility reason as mentioned by @tobealive I used tab in all my TS projects. Deno and Go comes with a native formatter that support to set tabs but some language don't offer native formatter configuration or didn't follow the same conventions. So, I think that each repo have to format with the simpliest or more common toolchain and follow their language conventions and use use standard conventions for others language include in their repo.

For the main (C) repo I assumed that as esbuild is needed for bundling bridge npm (and so package.json) can be a way to load hooks and tool chain but maybe python is a most common alternative.

If there is common style rules to follow, a language agnostic config or templates can be set up.
